typing/parmatch.ml

```diff
@@ -1291,68 +1291,108 @@ let print_pat pat =
-          (* first column of pss is made of variables only *)
-          let sub_witnesses = exhaust ext default (n-1) in
-          let q0 = Patterns.Head.to_omega_pattern q0 in
-          Seq.map (fun row -> q0::row) sub_witnesses
-      | { default; constrs } ->
-          let try_non_omega (p,pss) =
-            if is_absent_pat p then
-              Seq.empty
-            else
-              let sub_witnesses =
-                exhaust
-                  ext pss
-                  (List.length (simple_match_args p Patterns.Head.omega [])
-                   + n - 1)
-              in
-              let p = Patterns.Head.to_omega_pattern p in
-              Seq.map (set_args p) sub_witnesses
-          in
-          let try_omega () =
-            if full_match false constrs && not (should_extend ext constrs) then
-              Seq.empty
-            else
-              let sub_witnesses = exhaust ext default (n-1) in
-              match build_other ext constrs with
-              | exception Empty ->
-                  (* cannot occur, since constructors don't make
-                     a full signature *)
-                  fatal_error "Parmatch.exhaust"
-              | p ->
-                  Seq.map (fun tail -> p :: tail) sub_witnesses
-          in
-          (* Lazily compute witnesses for all constructor submatrices
-             (Some constr_mat) then the default submatrix (None).
-             Note that the call to [try_omega ()] is delayed to after
-             all constructor matrices have been traversed. *)
-          List.map (fun constr_mat -> Some constr_mat) constrs @ [None]
-          |> List.to_seq
-          |> Seq.flat_map
-            (function
-              | Some constr_mat -> try_non_omega constr_mat
-              | None -> try_omega ())
-    end
+| [(p :: ps)] -> exhaust_single_row ext p ps n
+| pss   -> specialize_and_exhaust ext pss n
+
+and exhaust_single_row ext p ps n =
+  (* Shortcut: in the single-row case p :: ps we know that all
+     counter-examples are either of the form
+       counter-example(p) :: omegas
+     or
+       p :: counter-examples(ps)
+
+     This is very interesting in the case where p contains
+     or-patterns, as the non-shortcut path below would do a separate
+     search for each constructor of the or-pattern, which can lead to
+     an exponential blowup on examples such as
+
+       | (A|B), (A|B), (A|B), (A|B) -> foo
+
+     Note that this shortcut also applies to examples such as
+
+       | A, A, A, A -> foo | (A|B), (A|B), (A|B), (A|B) -> bar
+
+     thanks to the [get_mins] preprocessing step which will drop the
+     first row (subsumed by the second). Code with this shape does
+     occur naturally when people want to avoid fragile pattern
+     matches: if A and B are the only two constructors, this is the
+     best way to make a non-fragile distinction between "all As" and
+     "at least one B".
+  *)
+  List.to_seq [Some p; None] |> Seq.flat_map
```

Comment:
It's unfortunate that this roundabout code is the easiest way of concatenating two lazy sequences. (It looks fine for now, but the existence of such code suggests that Seq is missing some features)

Reply:
There is `Seq.append : 'a t -> 'a t -> 'a t` since 4.11, but I wish to delay the computation of the second sequence. We would need a `Seq.append_lazy : 'a t -> (unit -> 'a t) -> 'a t`. The present code structure is slightly less clear, but (1) it mirrors the (equally unclear) structure of the non-single-row case, and (2) it made it very easy during my experiment to change iteration order. so there is something to it.